flecsi/exec/hpx/policy.hh

```diff
@@ -0,0 +1,215 @@
+
+template<typename R>
+struct reduction_helper {
+
+  template<typename T>
+  auto operator()(T const & lhs, T const & rhs) const {
+    return R::combine(lhs, rhs);
+  }
+};
+
+template<typename R,
+  task_processor_type_t ProcessorType,
+  typename Params,
+  typename F>
+auto
+delay_execution(prolog<ProcessorType> && prologue,
+  Params && params,
+  std::string && task_name,
+  F && f) {
+  return std::move(prologue).template delay_execution<R>(
+    std::move(params), std::move(task_name), std::forward<F>(f));
+}
+} // namespace detail
+
+template<auto & F, class Reduction, TaskAttributes Attributes, typename... Args>
+auto
+reduce_internal(Args &&... args) {
+  using Traits = util::function_traits<decltype(F)>;
+  using R = typename Traits::return_type;
+
+  // replace arguments in args, for example, field_reference -> accessor.
+  auto params = exec::detail::replace_arguments(
+    static_cast<typename Traits::arguments_type *>(nullptr),
+    std::forward<Args>(args)...);
+
+  auto task_name = util::symbol<F>();
+
+  // Now we have accessors, we need to bind the accessor to real memory
+  // for the data field. We also need to patch up default conversion
+  // from args to params, especially for the future<>.
+  // This is being achieved by creating the prolog<> instances below.
+
+  // Different kinds of task invocation with flecsi::execute():
+  // 1. domain_size is a std::monostate: a single task launch. On a single
+  //    rank (0) of our choice, we apply F to ARGS. Given the return type R of
+  //    F, we return a future<R, single>. The client can later call
+  //    future<>.get() on any rank to get the result. The value returned by
+  //    .get should be the same on all ranks, implying a Broadcast is needed.
+  // 2. Reduction is not void: a true reduction task.
+  //    We apply F to ARGS and reduce the return values with the Reduction
+  //    operation. We then put the reduced value into a future<R,
+  //    launch_type::single> and return it. Again, the client could call
+  //    .get() to get the (same) result on any rank, implying either an
+  //    Allreduce or Reduce/Broadcast is needed.
+  // 3. Reduction is void: an index launch. We apply F to ARGS on every rank,
+  //    each will return a value r_i. The value r_i will be put into the the
+  //    slot i in the future<R, index>. The client can call future<R,
+  //    index>.get(j) to get the return value on any rank j. This implies an
+  //    Allgather is needed.
+  const auto ds =
+    launch_size<Attributes, decltype(params)>(std::forward<Args>(args)...);
+  util::annotation::rguard<util::annotation::execute_task_user> ann{task_name};
+  if constexpr(std::is_same_v<decltype(ds), const std::monostate>) {
+    const bool root = flecsi::run::context::instance().process() == 0;
+
+    // single launch, only invoke the user task on the Root.
+    if constexpr(std::is_void_v<R>) {
+      if(root) {
+        prolog<mask_to_processor_type(Attributes)> bound_params(
```

Comment:
Would it be possible to hoist this out of the `if`s?  It seems that the only change would be that it happened on non-root processes for a single launch; if that's bad (or expensive), we could make it a `std::optional`.

The reason this is nice is it would allow `detail::delay_execution` to be a lambda:
```cpp
const auto delay=[&](auto &&f) {
  return std::move(bound_params).template delay_execution<R>(
     std::move(params), std::move(task_name), std::forward<F>(f));
};
```

Then calls to `delay([](auto &&params,std::string &&) {…})` might even be short enough to inline into the `return future<…>(…);`, saving another `std::move` each.  (`delay` could even have `std::apply(F, std::move(params));` as a default argument.)